File: BASIC/PowerOfTwo.vb
Imports System
Imports System.IO

Module PowerOfTwo
    Sub Main()
        Dim logPath As String = Path.Combine(Directory.GetCurrentDirectory(), "PowerOfTwo.log")
        Dim logFile As StreamWriter = File.CreateText(logPath)

        Dim header As String = "Computing powers of two (Big Integer)"
        Dim separator As String = "======================================"

        logFile.WriteLine(header)
        logFile.WriteLine(separator)
        logFile.WriteLine("")
        logFile.Flush()

        Console.WriteLine(header)
        Console.WriteLine(separator)
        Console.WriteLine("")

        Dim digits As Integer() = New Integer(10000) {}
        Dim count As Integer = 1
        Dim power As Long = 0

        digits(0) = 1

        While True
            Dim result As String = "2^" & power & " = " & ToString(digits, count)
            
            logFile.WriteLine(result)
            logFile.Flush()
            Console.WriteLine(result)

            MultiplyBy2(digits, count)
            power = power + 1
        End While

        logFile.Close()
    End Sub

    Sub MultiplyBy2(digits As Integer(), ByRef count As Integer)
        Dim i As Integer
        Dim product As Integer
        Dim carry As Integer = 0

        For i = 0 To count - 1
            product = digits(i) * 2 + carry
            digits(i) = product Mod 10
            carry = product \ 10
        Next i

        While carry > 0
            digits(count) = carry Mod 10
            carry = carry \ 10
            count = count + 1
        End While
    End Sub

    Function ToString(digits As Integer(), count As Integer) As String
        Dim result As String = ""
        Dim i As Integer

        For i = count - 1 To 0 Step -1
            result = result & digits(i).ToString()
        Next i

        Return result
    End Function
End Module

File: ObjectiveC/PowerOfTwo/PowerOfTwo.m
#import <Foundation/Foundation.h>

@interface BigInt : NSObject {
    NSMutableArray *digits;
}

- (instancetype)initWithInt:(int)value;
- (void)multiplyBy2;
- (NSString *)description;

@end

@implementation BigInt

- (instancetype)initWithInt:(int)value {
    self = [super init];
    if (self) {
        digits = [[NSMutableArray alloc] init];
        [digits addObject:@(value)];
    }
    return self;
}

- (void)multiplyBy2 {
    int carry = 0;
    for (NSInteger i = 0; i < [digits count]; i++) {
        int digit = [[digits objectAtIndex:i] intValue];
        int product = digit * 2 + carry;
        [digits replaceObjectAtIndex:i withObject:@(product % 10)];
        carry = product / 10;
    }
    
    while (carry > 0) {
        [digits addObject:@(carry % 10)];
        carry = carry / 10;
    }
}

- (NSString *)description {
    NSMutableString *result = [[NSMutableString alloc] init];
    for (NSInteger i = [digits count] - 1; i >= 0; i--) {
        [result appendString:[NSString stringWithFormat:@"%@", [digits objectAtIndex:i]]];
    }
    return result;
}

- (void)dealloc {
    [digits release];
    [super dealloc];
}

@end

int main(int argc, char *argv[]) {
    @autoreleasepool {
        NSString *logPath = [[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject] stringByAppendingPathComponent:@"PowerOfTwo.log"];
        
        NSFileManager *fileManager = [NSFileManager defaultManager];
        [fileManager removeItemAtPath:logPath error:nil];
        
        NSFileHandle *logFile = [NSFileHandle fileHandleForWritingAtPath:logPath];
        if (!logFile) {
            [[NSFileManager defaultManager] createFileAtPath:logPath contents:nil attributes:nil];
            logFile = [NSFileHandle fileHandleForWritingAtPath:logPath];
        }
        
        NSString *header = @"Computing powers of two (Big Integer)\n";
        NSString *separator = @"======================================\n\n";
        
        [logFile writeData:[header dataUsingEncoding:NSUTF8StringEncoding]];
        [logFile writeData:[separator dataUsingEncoding:NSUTF8StringEncoding]];
        
        printf("%s%s", [header UTF8String], [separator UTF8String]);
        
        BigInt *value = [[BigInt alloc] initWithInt:1];
        long power = 0;
        
        while (1) {
            NSString *result = [NSString stringWithFormat:@"2^%ld = %@\n", power, value];
            
            [logFile writeData:[result dataUsingEncoding:NSUTF8StringEncoding]];
            printf("%s", [result UTF8String]);
            fflush(stdout);
            
            [value multiplyBy2];
            power++;
        }
        
        [logFile closeFile];
        [value release];
    }
    
    return 0;
}
